Add declaration for Document.execCommand()

eemeli opened the pull request:
Adds a declaration for the document's [`execCommand()`](https://developer.mozilla.org/en-US/docs/Web/API/document.execCommand), copied from the already-implemented IE-only textRange's `execCommand()`.

gabelevi commented:
One of the annoying things about our test suite is that it's somewhat sensitive to line number changes in the flow libs. As you'll see from the travis result, the `union` test is failing. You can do something like

```Bash
flow check tests/union --all --strip-root > tests/union/union.exp
```

to update the expected errors.

eemeli commented:
That seems a bit... brittle. In any case, is build on OS X Yosemite supported? I have OCaml 4.02.1, installed via brew, but `make build-flow` fails for me with the following (first line abbreviated):

```
ocamlbuild  -no-links  -I src/stubs -I src/commands -I src/common -I src/embedded -I src/dts -I src/typing -I src/parser -I src/server -I src/parsing -I hack/utils -I hack/client -I hack/socket -I hack/server -I hack/stubs -I hack/typing -I hack/naming -I hack/parsing -I hack/deps -I hack/heap -I hack/globals -I hack/procs -I hack/search -I hack/hhi -I hack/dfind -I hack/third-party/avl -I hack/fsevents -I hack/fsnotify_darwin -lib unix -lib str -lflags "src/embedded/flowlib_elf.o hack/heap/hh_shared.o hack/utils/realpath.o hack/fsevents/fsevents_stubs.o hack/utils/nproc.o hack/hhi/hhi_elf.o hack/utils/get_build_id.gen.o hack/utils/get_build_id.o   -cclib -framework -cclib CoreServices -cclib -framework -cclib CoreFoundation -cclib -sectcreate -cclib __text -cclib flowlib -cclib /Users/eemeli/code/flow/bin/flowlib.tar.gz" src/flow.native
+ /usr/local/bin/ocamlc.opt -c -I hack/utils -I src/typing -I src/stubs -I src/server -I src/parsing -I src/parser -I src/embedded -I src/dts -I src/common -I src/commands -I hack/typing -I hack/stubs -I hack/socket -I hack/server -I hack/search -I hack/procs -I hack/parsing -I hack/naming -I hack/hhi -I hack/heap -I hack/globals -I hack/fsnotify_darwin -I hack/fsevents -I hack/dfind -I hack/deps -I hack/client -I hack/third-party/avl -o hack/utils/tmp.cmo hack/utils/tmp.ml
File "hack/utils/tmp.ml", line 17, characters 12-34:
Warning 3: deprecated: Filename.temp_dir_name
Use Filename.get_temp_dir_name instead
+ /usr/local/bin/ocamlopt.opt -c -I hack/utils -I src/typing -I src/stubs -I src/server -I src/parsing -I src/parser -I src/embedded -I src/dts -I src/common -I src/commands -I hack/typing -I hack/stubs -I hack/socket -I hack/server -I hack/search -I hack/procs -I hack/parsing -I hack/naming -I hack/hhi -I hack/heap -I hack/globals -I hack/fsnotify_darwin -I hack/fsevents -I hack/dfind -I hack/deps -I hack/client -I hack/third-party/avl -o hack/utils/tmp.cmx hack/utils/tmp.ml
File "hack/utils/tmp.ml", line 17, characters 12-34:
Warning 3: deprecated: Filename.temp_dir_name
Use Filename.get_temp_dir_name instead
+ /usr/local/bin/ocamlopt.opt src/embedded/flowlib_elf.o hack/heap/hh_shared.o hack/utils/realpath.o hack/fsevents/fsevents_stubs.o hack/utils/nproc.o hack/hhi/hhi_elf.o hack/utils/get_build_id.gen.o hack/utils/get_build_id.o -cclib -framework -cclib CoreServices -cclib -framework -cclib CoreFoundation -cclib -sectcreate -cclib __text -cclib flowlib -cclib /Users/eemeli/code/flow/bin/flowlib.tar.gz unix.cmxa str.cmxa hack/client/clientExceptions.cmx hack/utils/sys_utils.cmx hack/utils/path.cmx hack/client/clientLogCommand.cmx src/server/sysConfig.cmx hack/utils/tmp.cmx hack/socket/socket.cmx hack/utils/ident.cmx hack/utils/utils.cmx hack/utils/lock.cmx hack/client/clientUtils.cmx hack/utils/pidLog.cmx hack/client/clientStop.cmx hack/utils/build_id.cmx hack/utils/hh_json.cmx hack/utils/relative_path.cmx hack/utils/pos.cmx hack/utils/errors.cmx hack/server/serverMsg.cmx hack/client/hackClientStop.cmx hack/client/clientStart.cmx hack/stubs/eventLogger.cmx hack/utils/tty.cmx hack/client/clientBuild.cmx hack/client/clientEnv.cmx hack/client/clientProlog.cmx hack/client/clientStatus.cmx hack/client/clientCommand.cmx hack/utils/wwwroot.cmx hack/client/clientArgs.cmx hack/deps/fileInfo.cmx hack/deps/typing_deps.cmx hack/heap/prefix.cmx hack/heap/value.cmx hack/heap/sharedMem.cmx hack/naming/naming_special_names.cmx hack/parsing/namespace_env.cmx hack/parsing/ast.cmx hack/naming/nast.cmx hack/typing/typing_reason.cmx hack/typing/typing_defs.cmx hack/typing/typing_env.cmx hack/typing/typing_hooks.cmx hack/typing/typing_print.cmx hack/server/argumentInfoService.cmx hack/client/clientArgumentInfo.cmx hack/globals/autocomplete.cmx hack/globals/ide.cmx hack/naming/attributes.cmx hack/naming/naming_ast_helpers.cmx hack/naming/naming_hooks.cmx hack/parsing/namespaces.cmx hack/naming/naming.cmx hack/naming/nastVisitor.cmx hack/typing/nastCheck.cmx hack/typing/nast_terminality.cmx hack/typing/typeVisitor.cmx hack/typing/typing_expand.cmx hack/typing/typing_exts.cmx hack/typing/typing_generic.cmx hack/typing/typing_utils.cmx hack/typing/typing_instantiate.cmx hack/typing/typing_taccess.cmx hack/typing/typing_tdef.cmx hack/typing/typing_unification_env.cmx hack/typing/typing_unify.cmx hack/typing/typing_subtype.cmx hack/typing/typing_ops.cmx hack/typing/typing_suggest.cmx hack/typing/nastInitCheck.cmx hack/typing/typingEqualityCheck.cmx hack/typing/typing_alias.cmx hack/typing/typing_async.cmx hack/typing/typing_dynamic_yield.cmx hack/typing/typing_enum.cmx hack/typing/typing_extends.cmx hack/typing/typing_hint.cmx hack/typing/typing_lenv.cmx hack/typing/typing_variance.cmx hack/typing/typing.cmx hack/server/autocompleteService.cmx hack/client/clientAutocomplete.cmx hack/server/serverError.cmx hack/client/clientCheckStatus.cmx hack/client/colorFile.cmx hack/parsing/parser_heap.cmx hack/typing/coverage_level.cmx hack/client/clientColorFile.cmx hack/globals/serverConfig.cmx hack/heap/globalStorage.cmx hack/naming/naming_heap.cmx hack/parsing/lexer_hack.cmx hack/parsing/parser_hack.cmx hack/procs/bucket.cmx hack/utils/fork.cmx hack/utils/printSignal.cmx hack/procs/worker.cmx hack/procs/multiWorker.cmx hack/server/find.cmx hack/utils/config_file.cmx hack/server/serverArgs.cmx hack/server/serverEnv.cmx hack/typing/typing_inherit.cmx hack/typing/typing_decl.cmx hack/typing/typing_decl_service.cmx hack/typing/typing_check_service.cmx hack/server/serverIdeUtils.cmx hack/server/serverCoverageMetric.cmx hack/client/clientCoverageMetric.cmx hack/typing/typing_compare.cmx hack/server/findRefsService.cmx hack/server/serverFindRefs.cmx hack/client/clientFindRefs.cmx hack/server/methodJumps.cmx hack/client/clientMethodJumps.cmx hack/client/clientOutline.cmx hack/server/serverRefactor.cmx hack/client/clientRefactor.cmx hack/parsing/parsing_hooks.cmx hack/search/searchUtils.cmx hack/search/fuzzySearchService.cmx hack/utils/trie.cmx hack/search/trieSearchService.cmx hack/search/searchService.cmx hack/search/hackSearchService.cmx hack/client/clientSearch.cmx hack/client/clientTypeAtPos.cmx hack/server/fileOutline.cmx hack/server/inferAtPosService.cmx hack/server/serverArgumentInfo.cmx hack/server/serverColorFile.cmx hack/server/serverFileOutline.cmx hack/server/serverInferType.cmx hack/server/serverSearch.cmx hack/client/clientCheck.cmx hack/fsevents/fsevents.cmx hack/fsnotify_darwin/fsnotify.cmx hack/third-party/avl/monoidAvl.cmx hack/dfind/dfindEnv.cmx hack/dfind/dfindMaybe.cmx hack/dfind/dfindAddFile.cmx hack/dfind/dfindServer.cmx hack/dfind/dfindLib.cmx hack/hhi/hhi.cmx hack/parsing/parsing_service.cmx hack/server/serverDfind.cmx hack/server/serverEnvBuild.cmx hack/server/serverHealth.cmx hack/server/serverPeriodical.cmx hack/stubs/buildMain.cmx hack/server/serverFunctors.cmx src/common/flowConfig.cmx src/common/modes_js.cmx src/embedded/flowlib.cmx src/common/files_js.cmx src/commands/commandUtils.cmx src/parser/spider_monkey_ast.cmx src/common/reason_js.cmx src/parser/parse_error.cmx src/common/errors_js.cmx src/typing/constraint_js.cmx src/typing/flow_js.cmx src/typing/env_js.cmx src/typing/type_inference_hooks_js.cmx src/typing/autocomplete_js.cmx src/server/autocompleteService_js.cmx src/server/serverProt.cmx src/commands/autocompleteCommand.cmx src/commands/checkContentsCommand.cmx src/commands/configCommands.cmx src/dts/dts_ast.cmx src/dts/lexer_dts.cmx src/dts/parser_dts.cmx src/dts/printer_dts.cmx src/commands/convertCommand.cmx src/commands/findModuleCommand.cmx src/commands/getDefCommand.cmx src/commands/getImportersCommand.cmx src/commands/getImportsCommand.cmx src/commands/portCommand.cmx src/parser/lexer_flow.cmx src/parser/parser_flow.cmx src/parsing/parsing_service_js.cmx src/stubs/flowEventLogger.cmx src/typing/module_js.cmx src/typing/type_inference_js.cmx src/typing/comments_js.cmx src/typing/getDef_js.cmx src/typing/init_js.cmx src/typing/sort_js.cmx src/typing/types_js.cmx src/server/server.cmx src/commands/serverCommands.cmx src/commands/singleCommand.cmx src/commands/statusCommands.cmx src/commands/stopCommand.cmx src/commands/suggestCommand.cmx src/commands/typeAtPosCommand.cmx src/flow.cmx -o src/flow.native
Undefined symbols for architecture x86_64:
  "_caml_atom_table", referenced from:
      _caml_alloc in libasmrun.a(alloc.o)
      _caml_alloc_array in libasmrun.a(alloc.o)
      _caml_alloc_dummy in libasmrun.a(alloc.o)
      _caml_alloc_dummy_float in libasmrun.a(alloc.o)
      _intern_alloc in libasmrun.a(intern.o)
      _intern_rec in libasmrun.a(intern.o)
      _caml_make_float_vect in libasmrun.a(array.o)
      ...
  "_caml_code_area_end", referenced from:
      _segv_handler in libasmrun.a(signals_asm.o)
  "_caml_code_area_start", referenced from:
      _segv_handler in libasmrun.a(signals_asm.o)
  "start", referenced from:
     -u command line option
     (maybe you meant: _camlCamlinternalFormat__print_start_1150, _camlUtils__str_starts_with_1855 , _camlSys_utils__restart_1045 , _camlClientArgs__parse_restart_args_1310 , _camlClientArgs__parse_start_args_1309 , _caml_startup__data_end , _camlServerFunctors__start_1606 , _camlClientStart__should_start_1250 , _camlErrors__start_with_T_1581 , _caml_startup__data_begin , _camlWorker__start_worker_1047 , _camlDfindLib__start_1238 , _caml_young_start , _camlLexing__lexeme_start_1095 , _camlEventLogger__client_startup_1019 , _caml_startup__frametable , _caml_start_program , _camlCommandUtils__connect_with_autostart_1127 , _camlClientStart__start_server_1247 , _camlLexing__lexeme_start_p_1099 , _caml_heap_start , _caml_startup__code_begin , _camlCamlinternalFormat__parse_char_set_start_73212 , _caml_startup__code_end , _camlClientArgs__parse_start_env_1300 , _camlCommandUtils__start_flow_server_1075 )
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
File "caml_startup", line 1:
Error: Error during linking
Command exited with code 2.
Hint: Recursive traversal of subdirectories was not enabled for this build,
  as the working directory does not look like an ocamlbuild project (no
  '_tags' or 'myocamlbuild.ml' file). If you have modules in subdirectories,
  you should add the option "-r" or create an empty '_tags' file.
  
  To enable recursive traversal for some subdirectories only, you can use the
  following '_tags' file:
  
      true: -traverse
      <dir1> or <dir2>: traverse
      
Compilation unsuccessful after building 662 targets (0 cached) in 00:00:27.
make: *** [build-flow] Error 10
```

gabelevi commented:
I haven't upgrade to Yosemite yet out of laziness, but @jeffmo has and hasn't had any trouble. I also haven't heard any other reports of this yet...weird!

I googled a little bit for this, and all the occurrences I saw had to do with people writing C programs that call into OCaml and forget to startup OCaml first with some function call. Flow is an OCaml program that calls into C, so those answers don't seem to apply.

I wonder if maybe brew's pre-compiled version of ocaml has trouble on Yosemite? I have had problems with [homebrew and opam before](https://github.com/ocaml/opam/issues/1853), when the pre-built homebrew version was not compatible with the machine I was using. So maybe something like

```Bash
brew reinstall --build-from-source objective-caml
```

might solve your problem? Another way to install ocaml on a mac is using opam, like

```Bash
brew install opam
opam init --comp 4.02.1
```

Anyway, if all this is too much of a pain for you, I'm happy to fork your PR and fix the tests for you.

eemeli commented:
The issue was with my XCode command-line tools pointing to an older version (4.5.2) that I need for Android development; doing a `--build-from-source` for OCaml revealed this, and I fixed it by correcting the active developer directory using `xcode-select`.

I rebased my commit to include an update to the tests.

gabelevi commented:
Ah, great! Glad you found a fix! Sorry that you had to spend so much time on this and thanks for persevering!
